src/rgw/rgw_sal_dbstore.cc

```diff
@@ -757,6 +757,566 @@ namespace rgw::sal {
     return 0;
   }
 
+  int DBMultipartUpload::abort(const DoutPrefixProvider *dpp, CephContext *cct,
+				RGWObjectCtx *obj_ctx)
+  {
+    std::unique_ptr<rgw::sal::Object> meta_obj = get_meta_obj();
+    meta_obj->set_in_extra_data(true);
+    meta_obj->set_hash_source(mp_obj.get_key());
+    int ret;
+
+    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op = meta_obj->get_delete_op(obj_ctx);
+    del_op->params.bucket_owner = bucket->get_acl_owner();
+    del_op->params.versioning_status = 0;
+  
+    del_op->params.abortmp = true;
+
+    // Since the data objects are associated with meta obj till
+    // MultipartUpload::Complete() is done, removing the metadata obj
+    // should remove all the uploads so far.
+    ret = del_op->delete_obj(dpp, null_yield);
+    if (ret < 0) {
+      ldpp_dout(dpp, 20) << __func__ << ": del_op.delete_obj returned " <<
+        ret << dendl;
+    }
+    return (ret == -ENOENT) ? -ERR_NO_SUCH_UPLOAD : ret;
+  }
+
+  static string mp_ns = RGW_OBJ_NS_MULTIPART;
+
+  std::unique_ptr<rgw::sal::Object> DBMultipartUpload::get_meta_obj()
+  {
+    return bucket->get_object(rgw_obj_key(get_meta(), string(), mp_ns));
+  }
+
+  int DBMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y, RGWObjectCtx* obj_ctx, ACLOwner& owner, rgw_placement_rule& dest_placement, rgw::sal::Attrs& attrs)
+  {
+    int ret;
+    std::string oid = mp_obj.get_key();
+
+    char buf[33];
+    string tmp_obj_name;
+    std::unique_ptr<rgw::sal::Object> obj;
+    gen_rand_alphanumeric(store->ctx(), buf, sizeof(buf) - 1);
+    std::string upload_id = MULTIPART_UPLOAD_ID_PREFIX; /* v2 upload id */
+    upload_id.append(buf);
+
+    mp_obj.init(oid, upload_id);
+    tmp_obj_name = mp_obj.get_meta();
+
+    obj = get_meta_obj();
+    obj->set_in_extra_data(true);
+    obj->set_hash_source(oid);
+
+    DB::Object op_target(store->getDB(),
+	  		       obj->get_bucket()->get_info(),
+			       obj->get_obj());
+    DB::Object::Write obj_op(&op_target);
+
+    //op_target.set_versioning_disabled(true); /* no versioning for multipart meta 
+    obj_op.meta.owner = owner.get_id();
+    obj_op.meta.category = RGWObjCategory::MultiMeta;
+    obj_op.meta.flags = PUT_OBJ_CREATE_EXCL;
+    obj_op.meta.mtime = &mtime;
+
+    multipart_upload_info upload_info;
+    upload_info.dest_placement = dest_placement;
+
+    bufferlist bl;
+    encode(upload_info, bl);
+    obj_op.meta.data = &bl; 
+    ret = obj_op.prepare(dpp);
+    if (ret < 0)
+      return ret;
+    ret = obj_op.write_meta(dpp, bl.length(), bl.length(), attrs);
+
+    return ret;
+  }
+
+  int DBMultipartUpload::list_parts(const DoutPrefixProvider *dpp, CephContext *cct,
+				     int num_parts, int marker,
+				     int *next_marker, bool *truncated,
+				     bool assume_unsorted)
+  {
+    map<string, bufferlist> parts_map;
+    map<string, bufferlist>::iterator iter;
+
+    std::unique_ptr<rgw::sal::Object> obj = bucket->get_object(
```

Comment:
get_meta_obj() ?

Reply:
done. yes.. most of the code is common (as I chose to follow the Rados implementation) . I will check if I can wrap most of it as helper routines in  rgw_sal.cc

Reply:
I'm really kind of suspicious whether this code -should- closely resemble the rados code.  In particular, I've been worried about details like "head objects" and chunking crossing over, when there are basically no constraints selecting for those, coming from the dbstore itself.

Reply:
yes...there are no constrains from dbstore. It felt optimal to store object metadata & data separately (like in rados) and hence used same head and tail objects terms. 

Apart from that, the only thing common here is the way *MultipartUpload* processed (though it need not be). I will re-check on how much of it is actually needed for dbstore.